pkg/storage/stores/shipper/bloomshipper/store.go

```diff
@@ -0,0 +1,356 @@
+package bloomshipper
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path/filepath"
+	"sort"
+
+	"github.com/go-kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/storage"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/storage/chunk/client"
+	"github.com/grafana/loki/pkg/storage/config"
+)
+
+type Store interface {
+	SearchMetas(ctx context.Context, params MetaSearchParams) ([]Meta, error)
+	Fetcher(ts model.Time) *Fetcher
+	Stop()
+}
+
+var _ Client = &bloomStoreEntry{}
+var _ Store = &bloomStoreEntry{}
+
+type bloomStoreEntry struct {
+	start        model.Time
+	cfg          config.PeriodConfig
+	objectClient client.ObjectClient
+	bloomClient  Client
+	fetcher      *Fetcher
+}
+
+// SearchMetas implements store.
+func (b *bloomStoreEntry) SearchMetas(ctx context.Context, params MetaSearchParams) ([]Meta, error) {
+	var refs []MetaRef
+	tables := tablesForRange(b.cfg, params.Interval.Start, params.Interval.End)
+	for _, table := range tables {
+		prefix := filepath.Join(rootFolder, table, params.TenantID, metasFolder)
+		list, _, err := b.objectClient.List(ctx, prefix, "")
+		if err != nil {
+			return nil, fmt.Errorf("error listing metas under prefix [%s]: %w", prefix, err)
+		}
+		for _, object := range list {
+			metaRef, err := createMetaRef(object.Key, params.TenantID, table)
+
+			if err != nil {
+				return nil, err
+			}
+			if metaRef.MaxFingerprint < uint64(params.Keyspace.Min) || uint64(params.Keyspace.Max) < metaRef.MinFingerprint ||
+				metaRef.EndTimestamp.Before(params.Interval.Start) || metaRef.StartTimestamp.After(params.Interval.End) {
+				continue
+			}
+			refs = append(refs, metaRef)
+		}
+	}
+	return b.fetcher.FetchMetas(ctx, refs)
```

Comment:
Should this be done in a separate function call?

```go
refs, fetchers, err := store.SearchMetas(ctx, params)
for i := range refs {
    metas, err := fetchers[i].FetchMetas(ctx, refs[i])
}
```

Reply:
Yeah I'd keep these separate. Resolving Refs independently of fetching them allows us to do things without needing to download the underlying object. I think that's what you mean?

Reply:
Right, that's what I meant.